### PR TITLE
Add Mosaic Composer support for Ray

### PR DESCRIPTION
**Pull request by ilee300a**
In this PR, we provide initial commits for integrating Mosaic library with Ray. As Mosaic library provides algorithmic acceleration, providing further acceleration from the system side via Ray's distributed training, we can improve the speedup of training process.

Included in this PR is `MosaicTrainer` skeleton code, which currently can run a basic composer trainer without any logging.

### `MosaicTrainer` design
`MosaicTrainer` provides a wrapper for creating a worker loop that creates `composer.trainer.Trainer` and runs it.

`_mosaic_train_loop_per_worker`
→ Pre-worker training loop for Mosaic Composers
- sets environment variables for DDP
- gets dataset shards and converts them to an Iterable that can later be converted into `composer.core.DataSpec`
- create `composer.trainer.Trainer` and runs it

`_mosaic_utils.process_datasets`
→ process ray dataset shard into an iterable that can be converted into `composer.core.DataSpec`
- requires `batch_size : int` and `labels : List[str]` to be included in `trainer_init_config`
- Implementation
    - `RayDatasetMosaicIterable`
        - a wrapper class that provides iterator over ray dataset
    - `_mosaic_iterator`
        - iterator class for composer trainer that iterates over ray dataset.
        - uses `iter_torch_batches` to get each batch
        - requires that the dataset is mapped into pandas data frame with appropriate labels.
        - rearranges the DataFrame data into a list of each column data in the order the column labels are provided.
            - Enables the user to customize the data included in each batch.
            - Provides flexibility as the type and the number of data is clearly defined by the user.
            - compatible with composer’s training loop as the format of mini batch extraction training is `input, _ = batch` and `_, target = batch`

### Initial Experiments and Results
Experiment Setting(Omit all the hyperparameters unrelated to training speed):
- machine : m5.2xlarge
- Dataset : CIFAR 10
- Model : “resnet_56” from composer library
- Batch size : 64
- Num Epochs : 3
- No logging or callback

Experiment Result
- measured end-to-end training time (no eval phase)
- script example
    ```jsx
    start_time = time.perf_counter()
    trainer.fit()
    end_time = time.perf_counter()
    print(f"Total training time : {end_time - start_time:0.4f} seconds")
    ```

Trainer type | Total Time (s)
-- | --
Mosaic Composer | 764.2510
Mosaic Composer + Ray (5 workers) | 722.0651

### Future Steps
- Enable Checkpointing
  - Currently composer checkpointing works
  - Add support for Ray Checkpointing and session reports
- Enable Logging
  - `InMemoryLogger` does not work with `MosaicTrainer` at the moment. The code runs but the logger contains no data after training

### Other Notes
- Callbacks work with `MosaicTrainer`
- It may be useful to have `InMemoryLogger` by default and report that in `MosaicTrainer`
- Unit Test will be added once checkpointing has been added

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

**Comment by ilee300a** [comment]
Created a new PR [here](https://github.com/ray-project/ray/pull/28456)
